app/types/factory.py

```diff
@@ -0,0 +1,49 @@
+from abc import ABC, abstractmethod
+
+from sqlalchemy.ext.asyncio import AsyncSession
+
+
+class Factory(ABC):
+    def __init__(
+        self,
+        name: str,
+        depends_on: list[type["Factory"]],
+    ) -> None:
+        self.name = name
+        self.depends_on = depends_on
+
+    @abstractmethod
+    async def should_run(self, db: AsyncSession) -> bool:
+        """
+        Check if the factory should run.
+        This prevent duplicate runs of the same factory.
```

Comment:
Could you add a few words on the reason a factory may want to prevent its execution?